typescript/test/unit/value.ts

```diff
@@ -0,0 +1,317 @@
+          (messageString as unknown as JSONObject).valueField,
+          (messageBool as unknown as JSONObject).valueField,
+          (messageStruct as unknown as JSONObject).valueField,
+          {
+            listValue: {
+              values: [
+                {
+                  stringValue: 'test',
+                },
+              ],
+            },
+          },
+        ],
+      },
+    },
+  });
+  const jsonList = {
+    valueField: [
+      jsonNull.valueField,
+      jsonNumber.valueField,
+      jsonString.valueField,
+      jsonBool.valueField,
+      jsonStruct.valueField,
+      ['test'],
+    ],
+  };
+
+  it('serializes NullValue to proto3 JSON', () => {
+    const serialized = toProto3JSON(messageNull);
+    assert.deepStrictEqual(serialized, jsonNull);
+  });
+
+  it('deserializes NullValue from proto3 JSON', () => {
+    const deserialized = fromProto3JSON(MessageWithValue, jsonNull);
+    assert.deepStrictEqual(deserialized, messageNull);
+  });
+
+  it('serializes NumberValue to proto3 JSON', () => {
+    const serialized = toProto3JSON(messageNumber);
+    assert.deepStrictEqual(serialized, jsonNumber);
+  });
+
+  it('deserializes NumberValue from proto3 JSON', () => {
+    const deserialized = fromProto3JSON(MessageWithValue, jsonNumber);
+    assert.deepStrictEqual(deserialized, messageNumber);
+  });
+
+  it('serializes StringValue to proto3 JSON', () => {
+    const serialized = toProto3JSON(messageString);
+    assert.deepStrictEqual(serialized, jsonString);
+  });
+
+  it('deserializes StringValue from proto3 JSON', () => {
+    const deserialized = fromProto3JSON(MessageWithValue, jsonString);
+    assert.deepStrictEqual(deserialized, messageString);
+  });
+
+  it('serializes BoolValue to proto3 JSON', () => {
+    const serialized = toProto3JSON(messageBool);
+    assert.deepStrictEqual(serialized, jsonBool);
+  });
+
+  it('deserializes BoolValue from proto3 JSON', () => {
+    const deserialized = fromProto3JSON(MessageWithValue, jsonBool);
+    assert.deepStrictEqual(deserialized, messageBool);
+  });
+
+  it('serializes StructValue to proto3 JSON', () => {
+    const serialized = toProto3JSON(messageStruct);
+    assert.deepStrictEqual(serialized, jsonStruct);
+  });
+
+  it('deserializes StructValue from proto3 JSON', () => {
+    const deserialized = fromProto3JSON(MessageWithValue, jsonStruct);
+    assert.deepStrictEqual(deserialized, messageStruct);
+  });
+
+  it('serializes ListValue to proto3 JSON', () => {
+    const serialized = toProto3JSON(messageList);
+    assert.deepStrictEqual(serialized, jsonList);
+  });
+
+  it('deserializes ListValue from proto3 JSON', () => {
+    const deserialized = fromProto3JSON(MessageWithValue, jsonList);
+    assert.deepStrictEqual(deserialized, messageList);
+  });
+}
+
+function testGoogleProtobufStruct(root: protobuf.Root) {
+  const MessageWithStruct = root.lookupType('test.MessageWithStruct');
+  const message = MessageWithStruct.fromObject({
```

Comment:
So, if I'm understanding, we're going to be able to represent these deeply nested structures much more elegantly for users, e.g., the parser turns:

```js
    structField: {
      fields: {
        stringField: {
          stringValue: 'test',
        },
        numberField: {
          numberValue: 42,
        },
      },
    },
  });
```

Into:

```
 const json = {
    structField: {
      stringField: 'test',
      numberField: 42,
    },
  };
```

Reply:
That's right.  We are not yet enabling this for libraries, for now we just need to for the transport layer in REGAPIC since the HTTP endpoint expects this short form rather than the long one.

Enabling this for libraries is (1) breaking, (2) will require .d.ts updates to follow the short structure, so it's a separate story. But even now, aiplatform could start using this library instead of using their own conversion code.